### PR TITLE
Add new rounding modes to System.Math, System.MathF

### PR DESCRIPTION
**Pull request by hamish-rose**
resolves dotnet/corefx#31902

Adding new rounding modes to MidpointRounding for use with Math.Round and MathF.Round.

My unit tests are over on corefx at the moment. If I understand correctly I need my changes to be mirrored there by the bot and corefx be built with my coreclr version to run tests. 

See https://github.com/hamish-rose/corefx/blob/MidpointRoundingmodes%2331902/src/System.Runtime.Extensions/tests/System/Math.cs#L1469

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Decimal.cs [hunk 1]
```diff
@@ -617,7 +617,7 @@ private static decimal Round(ref decimal d, int decimals, MidpointRounding mode)
         {
             if ((uint)decimals > 28)
                 throw new ArgumentOutOfRangeException(nameof(decimals), SR.ArgumentOutOfRange_DecimalRound);
-            if ((uint)mode > (uint)MidpointRounding.AwayFromZero)
+            if ((uint)mode > (uint)MidpointRounding.ToPositiveInfinity)
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Decimal.cs]: `AwayFromZero` and `ToPositiveInfinity` are not the same and are not interchangeable.

File: src/System.Private.CoreLib/shared/System/Math.cs [hunk 2]
```diff
@@ -713,18 +713,40 @@ public static unsafe double Round(double value, int digits, MidpointRounding mod
...
+                else if (mode is MidpointRounding.ToZero)
                 {
-                    value = Round(value);
+                    value = Truncate(value);
+                }
+                else if (mode is MidpointRounding.ToNegativeInfinity)
+                {
+                    double fraction = ModF(value, &value);
+
+                    if (Sign(fraction) == -1)
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Math.cs]: This should only come into play for ties.

File: src/System.Private.CoreLib/shared/System/Decimal.cs (hunk 1 above)
Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Decimal.cs]: nevermind, this is the ArgumentException check, I misread the code.

File: src/System.Private.CoreLib/shared/System/Math.cs (hunk 2 above)
Comment (hamish-rose) [reply src/System.Private.CoreLib/shared/System/Math.cs]: Decimal.DecCalc.InternalRound with Truncate mode supplied (AKA ToZero if i'm not mistaken) currently behaves like this aswell.

e.g. these scenarios all pass..
```cs
[InlineData(11.0, 11, MidpointRounding.ToZero)]
[InlineData(11.4, 11, MidpointRounding.ToZero)]
[InlineData(11.5, 11, MidpointRounding.ToZero)]
[InlineData(11.6, 11, MidpointRounding.ToZero)]
[InlineData(-11.0, -11, MidpointRounding.ToZero)]
[InlineData(-11.4, -11, MidpointRounding.ToZero)]
[InlineData(-11.5, -11, MidpointRounding.ToZero)]
[InlineData(-11.6, -11, MidpointRounding.ToZero)]
public static void Round_Decimal_Modes(decimal x, decimal expected, MidpointRounding mode)
{
    Assert.Equal(expected, Math.Round(x, 0, mode));
    Assert.Equal(expected, decimal.Round(x, 0, mode));       
}
```
Agreed its not really midpoint rounding though, perhaps Math.Round(decimal, ..) could be implemented in Math.cs ?

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Math.cs]: Right. Sorry, I was thinking about the case where you are parsing a result.

Having comments explaining each case would be helpful. We should also just be able to use `Ceiling`, `Floor`, and `Truncate` (for `ToPositiveInfinity`, `ToNegativeInfinity`, and `Truncate`, respectively)

Comment (hamish-rose) [reply src/System.Private.CoreLib/shared/System/Math.cs]: > 
> 
> Right. Sorry, I was thinking about the case where you are parsing a result.
> 
> Having comments explaining each case would be helpful. We should also just be able to use `Ceiling`, `Floor`, and `Truncate` (for `ToPositiveInfinity`, `ToNegativeInfinity`, and `Truncate`, respectively)

Done, added some comments and now using Floor, Ceiling and Truncate.

File: src/System.Private.CoreLib/shared/System/MathF.cs [hunk 3]
```diff
@@ -183,18 +183,37 @@ public static unsafe float Round(float x, int digits, MidpointRounding mode)
 
                 x *= power10;
 
-                if (mode == MidpointRounding.AwayFromZero)
+                if (mode is MidpointRounding.ToEven)
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/MathF.cs]: would a switch statement be clearer?

**Comment by danmosemsft** [comment]
@hamish-rose I see you have tests in your corefx fork. What we typically do is put up the corefx PR with the tests at the same time as the coreclr PR with the implementation. Of course the build or tests fail in the corefx PR initially but it means that we can review the tests concurrently and we know before we merge the implementation that satisfcatory tests on it are ready.

File: src/System.Private.CoreLib/shared/System/MathF.cs (hunk 3 above)
Comment (hamish-rose) [reply src/System.Private.CoreLib/shared/System/MathF.cs]: We should need to select a mode (perhaps none) as the default case

**Comment by hamish-rose** [comment]
> 
> 
> @hamish-rose I see you have tests in your corefx fork. What we typically do is put up the corefx PR with the tests at the same time as the coreclr PR with the implementation. Of course the build or tests fail in the corefx PR initially but it means that we can review the tests concurrently and we know before we merge the implementation that satisfcatory tests on it are ready.

@danmosemsft i've raised https://github.com/dotnet/corefx/pull/33297

**Comment by tannergooding** [comment]
test this please

File: src/System.Private.CoreLib/shared/System/Math.cs [hunk 4]
```diff
@@ -713,18 +713,37 @@ public static unsafe double Round(double value, int digits, MidpointRounding mod
...
                     if (Abs(fraction) >= 0.5)
                     {
                         value += Sign(fraction);
                     }
                 }
-                else
+                else if (mode is MidpointRounding.ToZero)
                 {
-                    value = Round(value);
+                    // Directed rounding: Round toward zero, to nearest value above (positive numbers) or below (negative numbers)
```
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Math.cs]: I think the comment needs fixing, we won't be rounding above for positive numbers, but rather the nearest exact or the nearest just below (closer to zero)

**Comment by eerhardt** [comment]
>                 DecCalc.InternalRound(ref AsMutable(ref d), (uint)scale, (DecCalc.RoundingMode)mode);

Would it make sense to get rid of `DecCalc.RoundingMode` completely now and just use `public enum MidpointRounding` instead? That way:

1. We don't have 2 enums using different names to mean the same thing. It would aid in code readability.
2. We don't accidently get the values out of sync (since the way it exists now, the values need to match up for this to work).

---
Refers to: src/System.Private.CoreLib/shared/System/Decimal.cs:625 in 072f5b8. [](commit_id = 072f5b802c89d21183d575b053d2f817d42d3402, deletion_comment = False)

File: src/System.Private.CoreLib/shared/System/Math.cs [hunk 5]
```diff
@@ -713,18 +713,37 @@ public static unsafe double Round(double value, int digits, MidpointRounding mod
...
+                    value = Truncate(value);
+                }
+                else if (mode is MidpointRounding.ToNegativeInfinity)
+                {
+                    // Directed Rounding: Round down to the next value, toward −∞
```
Comment (eerhardt) [inline src/System.Private.CoreLib/shared/System/Math.cs]: I don't think we typically put non-standard symbols in source code, as some editors/viewers don't show these correctly.  Can we just put `negative infinity` instead?

File: src/System.Private.CoreLib/shared/System/MathF.cs (hunk 3 above)
Comment (hamish-rose) [reply src/System.Private.CoreLib/shared/System/MathF.cs]: I've updated the implementation in Math and MathF with a switch statement

**Comment by tannergooding** [comment]
@hamish-rose, sorry for the delay, this slipped off my radar.

Could you please either merge or rebase this PR against the current head of master? Once that is done (and the jobs all pass), we should be good to merge.

**Comment by hamish-rose** [comment]
@tannergooding I believe I've rebased onto master, working on the failing test now.

**Comment by krwq** [comment]
you can fix/restore most of the stuff with git reflog assuming you didn't remove local copy of the repository (copy the whole coreclr folder before just in case)

**Comment by hamish-rose** [comment]
On my local tracking branch I rebased it onto master, then when I pushed to my remote it 'merged' and now it a bit of a mess. 

Sorry about that, I'll give reflog a try.

**Comment by hamish-rose** [comment]
@tannergooding i've fixed the rebase now. The test that is failing is a corefx inner loop test that is obsolete now since we just have one enum for rounding mode. I think it can be removed once these changes are merged and mirrored in corefx

[Round_InvalidMidpointRounding_ThrowsArgumentException](https://github.com/dotnet/corefx/blob/08534ae1cf989696a0f1aac4a30a24e907983119/src/System.Runtime/tests/System/DecimalTests.cs#L928 )

**Comment by tannergooding** [comment]
For the failing CoreFX tests, we need to add an entry in the `CoreFX.issues.json` indicating it is outdated. You should just need to add the full name of the relevant test after the other `Math` test entries here: https://github.com/dotnet/coreclr/blob/master/tests/CoreFX/CoreFX.issues.json#L805

**Comment by tannergooding** [comment]
Thanks for the work here @hamish-rose

File: src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs [hunk 6]
```diff
@@ -2473,7 +2462,7 @@ internal static void InternalRound(ref DecCalc d, uint scale, RoundingMode mode)
                         ulong tmp = d.Low64;
                         if (tmp == 0)
                         {
-                            if (mode <= RoundingMode.Truncate)
+                            if (mode <= MidpointRounding.ToZero)
```
Comment (pentp) [inline src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs]: :exclamation: This is not correct - `Truncate` and `MidpointRounding.ToZero` are two completely different things: `Truncate(1.9m) == 1` while `MidpointRounding.ToZero(1.9m) == 2`

File: src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs [hunk 7]
```diff
@@ -2514,22 +2503,22 @@ internal static void InternalRound(ref DecCalc d, uint scale, RoundingMode mode)
...
-                else if (mode == RoundingMode.AwayFromZero)
+                else if (mode == MidpointRounding.AwayFromZero)
                 {
                     // Round away from zero at the mid point.
                     remainder <<= 1;
                     if (power > remainder)
                         goto done;
                 }
-                else if (mode == RoundingMode.Floor)
+                else if (mode == MidpointRounding.ToNegativeInfinity)
```
Comment (pentp) [inline src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs]: :exclamation: `Floor` and `MidpointRounding.ToNegativeInfinity` are not the same thing.

File: src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs [hunk 8]
```diff
@@ -2514,22 +2503,22 @@ internal static void InternalRound(ref DecCalc d, uint scale, RoundingMode mode)
...
                 {
                     // Round toward -infinity if we have chopped off a non-zero amount from a negative value.
                     if ((remainder | sticky) == 0 || !d.IsNegative)
                         goto done;
                 }
                 else
                 {
-                    Debug.Assert(mode == RoundingMode.Ceiling);
+                    Debug.Assert(mode == MidpointRounding.ToPositiveInfinity);
```
Comment (pentp) [inline src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs]: :exclamation: `Ceiling` and `MidpointRounding.ToPositiveInfinity` are not the same thing.

**Comment by pentp** [comment]
OK, the IEEE 754 spec actually talks about directed rounding (truncate/floor/ceiling), but then this API naming is definitely wrong - it's not __midpoint__ rounding.

**Comment by tannergooding** [comment]
@pentp, I've commented about your concerns on the original API proposal: https://github.com/dotnet/corefx/issues/31902.

We should continue the discussion there, to help keep it in one location.

**Pull request merged**